tee/src/lib.rs

```diff
@@ -0,0 +1,669 @@
+					Error::<T>::EnclaveAddressAlreadyRegisteredtoTheAccount
+				)
+			}
+
+			let (enclave_id, new_id) = Self::new_enclave_id()?;
+			// Needs to have enough money
+			let imbalance = T::Currency::withdraw(
+				&account,
+				T::EnclaveFee::get(),
+				WithdrawReasons::FEE,
+				KeepAlive,
+			)?;
+			T::FeesCollector::on_unbalanced(imbalance);
+
+			let enclave = Enclave::new(api_uri.clone(), enclave_address);
+
+			AccountEnclaveId::<T>::insert(account.clone(), enclave_id);
+			EnclaveData::<T>::insert(enclave_id, enclave);
+			EnclaveIdGenerator::<T>::put(new_id);
+
+			let reg_enclaves: BoundedVec<EnclaveId, T::MaxRegisteredEnclaves> =
+				<EnclaveRegistrationList<T>>::get();
+
+			let reg_enclaves = reg_enclaves
+				.try_mutate(|v| v.push(enclave_id))
+				.ok_or(Error::<T>::ErrorAddingToQueue)?;
+			<EnclaveRegistrationList<T>>::put(reg_enclaves);
+
+			Self::deposit_event(Event::EnclaveAdded { account, api_uri, enclave_id });
+			Ok(().into())
+		}
+
+		/// Removes an enclave from the system
+		/// Origin- operator account address
+		/// If the enclave is assigned, it will be placed in queue for tech committee approval
+		/// If enclave is not already assigned, he can exit without permission.
+		#[pallet::weight(T::WeightInfo::unregister_enclave())]
+		pub fn unregister_enclave(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			let account = ensure_signed_or_root(origin)?;
+			match account {
+				Some(account_id) => {
+					let enclave_id = AccountEnclaveId::<T>::get(account_id.clone())
+						.ok_or(Error::<T>::UnknownEnclaveOperatorAccount)?;
+
+					let unregistered_enclaves: BoundedVec<EnclaveId, T::MaxUnRegisteredEnclaves> =
+						<EnclaveUnregistrationList<T>>::get();
+
+					let reg_enclaves = <EnclaveRegistrationList<T>>::get();
+
+					if !reg_enclaves.contains(&enclave_id) {
+						EnclaveData::<T>::remove(enclave_id);
+						Self::deposit_event(Event::EnclaveUnregistered { account_id, enclave_id });
+					} else {
+						let unregistered_enclaves = unregistered_enclaves
+							.try_mutate(|v| v.push(enclave_id))
+							.ok_or(Error::<T>::ErrorAddingToQueue)?;
+						<EnclaveUnregistrationList<T>>::put(unregistered_enclaves);
+						Self::deposit_event(Event::MovedForUnregistration {
+							account_id,
+							enclave_id,
+						});
+					}
+				},
+				_ => {},
+			}
+
+			Ok(().into())
+		}
+
+		/// Update registration can be called only BEFORE assign_enclave.
+		#[pallet::weight(T::WeightInfo::update_registration())]
+		pub fn update_registration(
+			origin: OriginFor<T>,
+			enclave_address: Vec<u8>,
+			api_uri: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let account = ensure_signed(origin)?;
+
+			ensure!(api_uri.len() < T::MaxUriLen::get().into(), Error::<T>::UriTooLong);
+			ensure!(api_uri.len() > T::MinUriLen::get().into(), Error::<T>::UriTooShort);
+
+			let enclave_id =
+				AccountEnclaveId::<T>::get(&account).ok_or(Error::<T>::NotEnclaveOwner)?;
+
+			ensure!(
```

Comment:
Here, we better check the existence in the registration list

Reply:
@Leouarz `EnclaveRegistrationList` does not contain the account, Its is a list of enclaveIds. 

Since we are keeping a record on AccountEnclaveId, its convenient to receive the enclaveId by account